renderdoc/driver/ihv/amd/amd_counters.cpp

```diff
@@ -227,140 +414,374 @@ CounterDescription AMDCounters::InternalGetCounterDescription(uint32_t internalI
 {
-  m_pGPUPerfAPI->GPA_EndSession();
+  auto status = m_pGPUPerfAPI->GPA_EndSession();
+  if (AMD_FAILED(status))
+  {
+    GPA_LoggingCallback(GPA_LOGGING_ERROR, FormatErrMessage("End session.", status).c_str());
+  }
+}
+
+std::vector<CounterResult> AMDCounters::GetCounterData(
+  uint32_t sessionID,
+  uint32_t maxSampleIndex,
+  const std::vector<uint32_t> &eventIDs,
+  const std::vector<GPUCounter> &counters
+)
+{
+  std::vector<CounterResult> ret;
+
+  bool isReady = false;
+
+  const uint32_t timeout = 10000; // ms
+
+  auto startTime = std::chrono::high_resolution_clock::now();
+
+  do
+  {
+    isReady = IsSessionReady(sessionID);
+    if (!isReady)
+    {
+      std::this_thread::sleep_for(std::chrono::milliseconds(0));
+
+      auto endTime = std::chrono::high_resolution_clock::now();
+      std::chrono::duration<double, std::milli> elapsedTime = endTime - startTime;
+      if (elapsedTime.count() > timeout)
+      {
+        GPA_LoggingCallback(GPA_LOGGING_ERROR, "GetCounterData failed due to elapsed timeout.");
+        return std::move(ret);
+      }
+    }
+  } while (!isReady);
+
+  for (uint32_t s = 0; s < maxSampleIndex; s++)
+  {
+    for (size_t c = 0; c < counters.size(); c++)
+    {
+      const CounterDescription desc = GetCounterDescription(counters[c]);
+
+      switch (desc.resultType)
+      {
+      case CompType::UInt:
+      {
+        if (desc.resultByteWidth == sizeof(uint32_t))
+        {
+          uint32_t value = GetSampleUint32(sessionID, s, counters[c]);
+
+          if (desc.unit == CounterUnit::Percentage)
+          {
+            value = RDCCLAMP(value, 0U, 100U);
+          }
+
+          ret.push_back(CounterResult(eventIDs[s], counters[c], value));
+        }
+        else if (desc.resultByteWidth == sizeof(uint64_t))
+        {
+          uint64_t value = GetSampleUint64(sessionID, s, counters[c]);
+
+          if (desc.unit == CounterUnit::Percentage)
+          {
+            value = RDCCLAMP(value, (uint64_t)0, (uint64_t)100);
+          }
+
+          ret.push_back(CounterResult(eventIDs[s], counters[c], value));
+        }
+        else
+        {
+          RDCERR("Unexpected byte width %u", desc.resultByteWidth);
+        }
+      }
+      break;
+      case CompType::Float:
+      {
+        float value = GetSampleFloat32(sessionID, s, counters[c]);
+
+        if (desc.unit == CounterUnit::Percentage)
+        {
+          value = RDCCLAMP(value, 0.0f, 100.0f);
+        }
+
+        ret.push_back(CounterResult(eventIDs[s], counters[c], value));
+      }
+      break;
+      case CompType::Double:
+      {
+        double value = GetSampleFloat64(sessionID, s, counters[c]);
+
+        if (desc.unit == CounterUnit::Percentage)
+        {
+          value = RDCCLAMP(value, 0.0, 100.0);
+        }
+
+        ret.push_back(CounterResult(eventIDs[s], counters[c], value));
+      }
+      break;
+      default: RDCASSERT(0); break;
+      };
+    }
+  }
+
+  return std::move(ret);
```

Comment:
I think as discussed in `EnumerateCounters` all uses of `std::move` like this should be removed, the functions aren't returning rvalue references so the copy will still happen, leaving this confusing/misleading about what it does.

Reply:
Previously fixed